src/System.Management.Automation/engine/parser/Parser.cs

```diff
@@ -4560,6 +4560,19 @@ private MemberAst ClassMemberRule(string className, out List<Ast> astsOnError)
             return null;
         }
 
+        private bool CanBeSimpleNameToken(ref Token token)
+        {
+            if (token.Kind == TokenKind.Identifier
+                || token.Kind == TokenKind.DynamicKeyword
+                || token.TokenFlags.HasFlag(TokenFlags.Keyword))
+            {
+                token.TokenFlags = TokenFlags.None;
+                return true;
+            }
+
+            return false;
+        }
```

Comment:
When a keyword is tokenized as a simple name, the token kind should be `Identifier` and the `TokenFlags` should be set appropriately (in our case here, should be `MemberName` I believe)

The following is an example where the class name is parsed using `SimpleNameRule`. Take a look at the second `class` token.
```
$tokens = $null
$s = [System.Management.Automation.Language.Parser]::ParseInput("class class {  }", [ref]$tokens, [ref]$null)
PS:20> $tokens

Text       : class
TokenFlags : Keyword
Kind       : Class
HasError   : False
Extent     : class

Text       : class
TokenFlags : TypeName
Kind       : Identifier
HasError   : False
Extent     : class
...
```
So here, the utility method probably should be `TryUseAsSimpleName(Token old, out Token new)`, where `new` should be a new token that has `Kind = Identifier` and `TokenFlags = MemberName`. But then, the new token needs to be put in the `Tokenizer.TokenList` and replace the old token (see `Tokenizer.ReplaceSavedTokens`) ... now it would be more or less the same as `Resync` in terms of the overhead. Given that, maybe it makes sense to just `Resync` and call the `SameNameRule`?

Reply:
Hmm, maybe?

When I checked the difference in token kind and flags here, I noted that there appear to be no flags set at all for method names. Is MemberName used for properties instead, perhaps? :thinking:

Reply:
I'm moving away from the idea to `Resync` and use `SimpleNameRule` here. The [`FunctionDeclarationRule`](https://github.com/PowerShell/PowerShell/blob/4ae04335a498ccbc8f3a2eb16bbd2dc54f0956e1/src/System.Management.Automation/engine/parser/Parser.cs#L5234-L5272) accepts the token as is for the function name.

Note that for function declaration, a wider range of tokens (_more than just simple names, e.g. number token is also valid_) are considered valid and that's different from what a class method name can take, but I guess it's OK to not have the class name token shown up as `Identifier` kind. We can go back to the `Resync` and `SimpleNameRule` if this becomes a problem.